### PR TITLE
[Chips] Adds support for reading contentHorizontalAlignment in MDCChipView.

### LINKED ISSUE #4277: [Chips] Support UIControlContentHorizontalAlignment
## Feature Request/Suggestion

As a UIControl subclass, MDCChipView should make use of `UIControlContentHorizontalAlignment` to allow clients to customize the position of the content within the Chip.

## Additional context
There is currently an undesirable effect when increasing the width of an MDCChipView. It will center the title but leave the imageView/accessoryView on the edges. We have internal clients that would like all of the contents to be centered within the ChipView, with appropriate layout/padding between the subviews.

chip-sizing-mdcdragon

Related internal issue: b/80277575

### PR DESCRIPTION
**Pull request by compositeprimes**
The only supported values so far are Fill (the default now) and Center (the new value).

Also adds an example to the MDCDragon app:
screen shot 2018-06-05 at 2 03 17 pm

Closes #4277

### CODE REVIEW

**Comment by compositeprimes** [comment]
Trying again for the CLA check

**Comment by jverkoey** [comment]
Thanks Joe! Can you add a screenshot to the PR description? You can drag a png into the PR description text field and GitHub will upload the image for you.

**Comment by compositeprimes** [comment]
Added an image in the PR description

File: components/Chips/src/MDCChipView.m
Comment (jverkoey) [inline components/Chips/src/MDCChipView.m]: Can we do an early exit?

```
if (alignment != UIControlContentHorizontalAlignmentCenter) {
  return contentRect;
}

.... center alignment behavior
```

**Review by jverkoey — APPROVED** [review]

File: components/Chips/src/MDCChipView.m
Comment (compositeprimes) [reply components/Chips/src/MDCChipView.m]: Good idea, done

**Comment by jverkoey** [comment]
Hey @compositeprimes, I left a comment on the internal bug to clarify one thing before we proceed with this, but otherwise this PR looks good functionality-wise.

**Comment by yarneo** [comment]
@jverkoey do you think this is ready to merge?

File: components/Chips/src/MDCChipView.h
Comment (romoore) [inline components/Chips/src/MDCChipView.h]: Nit: I think it's better to reference the `contentHorizontalAlignment` property directly here.  Using Fill and Center instead of Default and Centered would also help clarify.

**Comment by jverkoey** [comment]
No, I’d like to get confirmation on the internal bug first.

**Comment by romoore** [comment]
Has this been reviewed by the designers? I don't see this type of layout ("stretched chips") at [material.io](https://material.io/design/components/chips.html) so I want to make sure we're building something we expect to support long-term.

File: components/Chips/src/MDCChipView.m
Comment (romoore) [inline components/Chips/src/MDCChipView.m:663]: Please use property access instead of ivars directly.

s/_titleLabel/self.titleLabel/

and others below

**Comment by jverkoey** [comment]
Before pulling in design I’d like to get confirmation from the internal bug filer on my question. We may not need this PR.

**Review by romoore — DISMISSED** [review]
Could you please add unit tests for the changes?

File: components/Chips/src/MDCChipView.m
Comment (compositeprimes) [reply components/Chips/src/MDCChipView.m:663]: There are 15 references to _titleLabel in this class, and only 3 to self.titleLabel. I'm going to go for consistency here.

Comment (romoore) [reply components/Chips/src/MDCChipView.m:663]: I think in this case, following best practices should trump consistency. I wouldn't mind switching the file over or just being consistent within the new code (preferring properties).

From Apple's guidelines:

> Although **it’s best practice for an object to access its own properties using accessor methods or dot syntax**, it’s possible to access the instance variable directly from any of the instance methods in a class implementation. ...
In general, you should use accessor methods or dot syntax for property access even if you’re accessing an object’s properties from within its own implementation, in which case you should use self... **The exception to this rule is when writing initialization, deallocation or custom accessor methods**, as described later in this section. (Emphasis added)
> &mdash; [Encapsulating Data (Accessed: June 6, 2018)](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/EncapsulatingData/EncapsulatingData.html)

Comment (compositeprimes) [reply components/Chips/src/MDCChipView.m:663]: The style guide that MDC says to use is the Google one (https://github.com/google/styleguide/blob/gh-pages/objcguide.md), which currently does not favor one or the other. But it does say this:
> When the style guide allows multiple options it is preferable to pick one option over mixed usage of multiple options. Using one style consistently throughout a codebase lets engineers focus on other (more important) issues. Consistency also enables better automation because consistent code allows more efficient development and operation of tools that format or refactor code. In many cases, rules that are attributed to "Be Consistent" boil down to "Just pick one and stop worrying about it"; the potential value of allowing flexibility on these points is outweighed by the cost of having people argue over them.

Comment (romoore) [reply components/Chips/src/MDCChipView.m:663]: Understood and I'm aware of the consistency guidelines as well.  I agree that this shouldn't be debated, I was just hoping we could clean up the code a bit since it was being worked on.

Comment (compositeprimes) [reply components/Chips/src/MDCChipView.m:663]: I understand your desire for it to get cleaned up, but as an example, while writing this code, I wasn't sure whether to use _titleLabel or self.titleLabel (and some for other ivars vs properties). I looked through the code and found more usages of the ivar, so I used that. I'd prefer to do the cleanup in one fell swoop as to not add to the confusion (i.e. more people like me, editing code here and being unsure which to use).

**Comment by jverkoey** [comment]
I've resolved the internal blocking conversation and this PR is good to go. @romoore is the PR still in a needs changes state?

**Comment by jverkoey** [comment]
@compositeprimes it looks like your commit was signed with a default email. You can fix this by rewriting the commit history like so:

```
git commit --amend --author="compositeprimes <your.email@address.com>"
git push --force
```

**Comment by romoore** [comment]
@jverkoey I'm still waiting on feedback/changes related to:
1. Unit tests (please!)
2. The documentation of the horizontal content alignment.
3. You said there was an internal issue to get designer approval. Could you please add a link?

File: components/Chips/src/MDCChipView.h
Comment (compositeprimes) [reply components/Chips/src/MDCChipView.h]: Done.

**Comment by compositeprimes** [comment]
@romoore What type of UnitTest are you expecting, exactly? UnitTesting layout is extremely brittle, and basically creates change-detector tests (which are usually frowned upon). There also don't seem to be any existing layout-checking UnitTests. Other than that type, I'm not sure what there is to UnitTest here

**Comment by jverkoey** [comment]
@romoore Can you file a follow-up issue outlining the desired unit tests?

**Comment by romoore** [comment]
@compositeprimes I think #4371 has a good example of unit tests for layout that shouldn't be brittle. I think something similar can be done here - ensuring that "centered" places all of the elements away from the edges while "fill" keeps the leading/trailing images near their respective edges.

**Comment by compositeprimes** [comment]
Added tests.

**Comment by ianegordon** [comment]
Thank you @compositeprimes for adding the tests!
@romoore would you approve / merge if everything looks good?

**Comment by romoore** [comment]
Looks good to me, thanks for the PR (and tests!) @compositeprimes .

**Comment by compositeprimes** [comment]
Thanks all, going to merge this in now :)

**Comment by compositeprimes** [comment]
Oh nevermind, you already did that Rob–thanks!

**Pull request merged**
